harness/determined/core/_profiler.py

```diff
@@ -0,0 +1,412 @@
+
+from determined import core
+from determined.common import api, constants
+from determined.common.api import bindings
+
+try:
+    import pynvml
+except ImportError:
+    pynvml = None
+
+logger = logging.getLogger("determined.core")
+
+
+class ProfilerContext:
+    """
+    ``ProfilerContext`` gives access to the system profiling feature within Determined.
+
+    It is responsible for collecting system metrics at specified time intervals and reporting
+    them to the master.
+    """
+
+    def __init__(
+        self,
+        session: api.Session,
+        agent_id: str,
+        trial_id: int,
+        run_id: int,
+        distributed: core.DistributedContext,
+    ) -> None:
+        self._session = session
+        self._agent_id = agent_id
+        self._trial_id = trial_id
+        self._run_id = run_id
+        self._distributed = distributed
+        self._on = False
+
+        self._collector = None
+        self._shipper = None
+
+    def on(self, interval: int = 1) -> None:
+        """
+        Turns system profiling functionality on.
+
+        This method creates two threads, one that collects system metrics at specified time
+        intervals, and another that ships them to the master.
+
+        .. note::
+
+            This method is idempotent: if profiling is already on, this method is effectively a
+            no-op.
+
+        Arguments:
+            interval: time (in seconds) between each metric collection.
+
+        """
+        if self._on:
+            return
+
+        # Only enable on local chief workers.
+        if self._distributed.local_rank != 0:
+            return
+
+        logger.info("Starting system metrics profiling.")
+
+        metrics_queue = queue.Queue()
+        self._collector = _Collector(
+            metrics_queue=metrics_queue,
+            collection_interval=interval,
+        )
+        self._shipper = _Shipper(
+            session=self._session,
+            agent_id=self._agent_id,
+            trial_id=self._trial_id,
+            run_id=self._run_id,
+            metrics_queue=metrics_queue,
+        )
+
+        self._collector.start()
+        self._shipper.start()
+        self._on = True
+
+    def off(self) -> None:
+        """
+        Turns off profiling.
+
+        .. note::
+
+            This method is idempotent: if profiling is already off, this method is effectively a
+            no-op.
+        """
+        if not self._on:
+            return
+
+        logger.info("Stopping system metrics profiling.")
+        self.close()
+        self._on = False
+
+    def close(self) -> None:
+        if self._collector:
+            self._collector.stop()
+        if self._shipper:
+            self._shipper.stop()
+
+
+class DummyProfilerContext(ProfilerContext):
+    def on(self, interval: int = 1) -> None:
+        pass
+
+    def off(self) -> None:
+        pass
+
+    def close(self) -> None:
+        pass
+
+
+class _MetricGroupCollector(metaclass=abc.ABCMeta):
```

Comment:
What are your thoughts about making this class public and subclassable by users?

Reply:
not possible. maybe at a way later point in time, but at this point no way because backend APIs only recognize certain groups, and web UI has even harder dependencies on which groups it can render appropriately.

Reply:
Thanks. Works for me!